### PR TITLE
Cleanup and improve the unified slash syntax

### PR DESCRIPTION
**Pull request by dwijnand**

### CODE REVIEW

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3596/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3596/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

File: internal/util-collection/src/main/scala/sbt/internal/util/Attributes.scala [hunk 1]
```diff
@@ -92,10 +92,8 @@ object AttributeKey {
       rank0: Int
   )(implicit mf: Manifest[T], ojw: OptJsonWriter[T]): AttributeKey[T] =
     new SharedAttributeKey[T] {
-      require(name.headOption match {
-        case Some(c) => c.isLower
-        case None    => false
-      }, s"A named attribute key must start with a lowercase letter: $name")
+      require(name.headOption.fold(false)(_.isLower),
```
Comment (eed3si9n) [inline internal/util-collection/src/main/scala/sbt/internal/util/Attributes.scala]: Pattern match is the most readable and most performant ("probably at least 10 times as fast as the alternatives" according to Martin).

File: main/src/main/scala/sbt/SlashSyntax.scala [hunk 2]
```diff
@@ -23,126 +22,75 @@ import sbt.librarymanagement.Configuration
...
+   * if we're task-scoping.
+   *
+   * @param scope the built-up scope
+   * @param key a given key
+   * @tparam K the type of the given key, necessary to type "materialize"
    */
-  final class TaskKeyThunk[A](base: Scope, key: TaskKey[A]) {
-    private[sbt] def materialize: TaskKey[A] = key in base
-    private[sbt] def rescope: RichScope = new RichScope(base in key.key)
+  final class ScopeAndKey[K <: Key[K]](scope: Scope, key: K) {
```
Comment (eed3si9n) [inline main/src/main/scala/sbt/SlashSyntax.scala:91]: Looks more elegant than my case-by-case approach.

**Review by eed3si9n — APPROVED** [review]
Added a minor nitpick on pattern matching.
Overall LGTM

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3596/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3596/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Pull request merged**